src/java.base/share/classes/java/util/Set.java

```diff
@@ -728,10 +728,25 @@ static <E> Set<E> of(E... elements) {
     static <E> Set<E> copyOf(Collection<? extends E> coll) {
         if (coll instanceof ImmutableCollections.AbstractImmutableSet) {
             return (Set<E>)coll;
-        } else if (coll.isEmpty()) { // Implicit nullcheck of coll
+        }
+        var it = coll.iterator();   // Implicit nullcheck of coll
+        if (!it.hasNext()) {
             return Set.of();
-        } else {
-            return (Set<E>)Set.of(new HashSet<>(coll).toArray());
         }
+        E e0 = it.next();
+        if (!it.hasNext()) {
+            return Set.of(e0);
+        }
+        E e1 = it.next();
+        if (!it.hasNext()) {
+            return Set.of(e0, e1);
```

Comment:
Bad change, doesn't handle e0.equals(e1), and this is getting beyond the original issue

Reply:
> Bad change, doesn't handle e0.equals(e1), and this is getting beyond the original issue

`Set.of(E, E)` handles duplication. If `coll` is `{Jumbo,Regular}EnumSetCompatible`, then we don't have to create the `HashSet`.

Reply:
no, Set.copyOf explicitly allows duplications, unlike Set.of. You need to read the specs so you know what you are doing.